src/GraphQl/Type/FieldsBuilder.php

```diff
@@ -0,0 +1,371 @@
+                if (!isset($value['#name'])) {
+                    $name = (false === $pos = strrpos($original, '[')) ? $original : substr($original, 0, (int) $pos);
+                    $value['#name'] = ($resourceMetadata ? $resourceMetadata->getShortName() : '').'Filter_'.strtr($name, ['[' => '_', ']' => '', '.' => '__']);
+                }
+            }
+
+            $args[$key] = $value;
+        }
+
+        return $args;
+    }
+
+    private function convertFilterArgsToTypes(array $args): array
+    {
+        foreach ($args as $key => $value) {
+            if (strpos($key, '.')) {
+                // Declare relations/nested fields in a GraphQL compatible syntax.
+                $args[str_replace('.', '_', $key)] = $value;
+                unset($args[$key]);
+            }
+        }
+
+        foreach ($args as $key => $value) {
+            if (!\is_array($value) || !isset($value['#name'])) {
+                continue;
+            }
+
+            $name = $value['#name'];
+
+            if ($this->typesContainer->has($name)) {
+                $args[$key] = $this->typesContainer->get($name);
+                continue;
+            }
+
+            unset($value['#name']);
+
+            $filterArgType = new InputObjectType([
+                'name' => $name,
+                'fields' => $this->convertFilterArgsToTypes($value),
+            ]);
+
+            $this->typesContainer->set($name, $filterArgType);
+
+            $args[$key] = $filterArgType;
+        }
+
+        return $args;
+    }
+
+    /**
+     * Converts a built-in type to its GraphQL equivalent.
+     *
+     * @throws InvalidTypeException
+     */
+    private function convertType(Type $type, bool $input, ?string $queryName, ?string $mutationName, string $resourceClass, ?string $property, int $depth)
+    {
+        $graphqlType = $this->typeConverter->convertType($type, $input, $queryName, $mutationName, $resourceClass, $property, $depth);
+
+        if (null === $graphqlType) {
+            throw new InvalidTypeException(sprintf('The type "%s" is not supported.', $type->getBuiltinType()));
+        }
+
+        if (\is_string($graphqlType)) {
+            if (!$this->typesContainer->has($graphqlType)) {
+                throw new InvalidTypeException(sprintf('The GraphQL type %s is not valid. Valid types are: %s. Have you registered this type by implementing %s?', $graphqlType, implode(', ', array_keys($this->typesContainer->all())), TypeInterface::class));
+            }
+
+            $graphqlType = $this->typesContainer->get($graphqlType);
+        }
+
+        if ($this->typeBuilder->isCollection($type)) {
+            return $this->paginationEnabled && !$input ? $this->typeBuilder->getResourcePaginatedCollectionType($graphqlType) : GraphQLType::listOf($graphqlType);
+        }
+
+        return $type->isNullable() || (null !== $mutationName && 'update' === $mutationName) ? $graphqlType : GraphQLType::nonNull($graphqlType);
```

Comment:
Because of this condition I can't customize GraphQLlType for class/entity properties.

Even if I decorate TypeConverter ([called above at line 351](https://github.com/alanpoulain/core/blob/6e13e6628126027cfe3bbc60b3d743669ccafcd1/src/GraphQl/Type/FieldsBuilder.php#L351)) and return exactly the type I need, it is forced to `nonNull` by this condition (even if I don't need `nonNull`) and I can't do nothing about it (because it's inside a private method).

The only solution was to fork API Platform and [replace this condition with just `return $graphqlType;`](https://github.com/bagrinsergiu/api-platform-core/commit/17ad2ad1c83fe3a1e22226f5041a8d0429dcbf77). Also this condition can't be removed because it would introduce breaking changes for existing projects _(all auto generated entity property types will be [`String` instead of `String!`](https://github.com/alanpoulain/core/blob/6e13e6628126027cfe3bbc60b3d743669ccafcd1/tests/GraphQl/Type/FieldsBuilderTest.php#L352-L361))_.

I would like this condition to be moved in a public method which is part of interface, so I can decorate this service and change method behavior to `return $graphqlType;`.

Reply:
If your property type is nullable (for instance with typehint or PHPDoc), the type is not wrapped. Which is your use case?

Reply:
1. I created an annotation where I specify explicitly my type.
2. I created a compiler pass to read all classes, extract the annotated properties and cache the result as a service argument.
3. I decorated TypeConverter to return the annotated explicit class property type by using the cache from previous step.
    ```php
    <?php
    
    declare(strict_types=1);
    
    namespace App\Type\GraphQL;
    
    use ApiPlatform\Core\GraphQl\Type\TypeConverterInterface;
    use App\Service\GraphQL\GraphQLTypeAnnotationCache;
    use GraphQL\Type\Definition\Type as GraphQLType;
    
    class TypeConverter implements TypeConverterInterface
    {
        private TypeConverterInterface $decorated;
        private GraphQLTypeAnnotationCache $annotationCache;
    
        public function __construct(
            TypeConverterInterface $decorated,
            GraphQLTypeAnnotationCache $annotationCache
        ) {
            $this->decorated = $decorated;
            $this->annotationCache = $annotationCache;
        }
    
        public function convertType(
            Type $type,
            bool $input,
            ?string $queryName,
            ?string $mutationName,
            ?string $subscriptionName,
            string $resourceClass,
            string $rootResource,
            ?string $property,
            int $depth
        ) {
            // required to be called before annotation cache getter
            //   because entity types must be defined before use
            $defaultType = $this->decorated->convertType(
                $type,
                $input,
                $queryName,
                $mutationName,
                $subscriptionName,
                $rootResource,
                $rootResource,
                $property,
                $depth
            );
    
            if ($property && ($customTypeName = $this->annotationCache->getGraphQLType($rootResource, $property))) {
                return $this->resolveType($customTypeName);
            }
    
            return $defaultType;
        }

        public function resolveType(string $type): ?GraphQLType
        {
            return $this->decorated->resolveType($type);
        }
    }
    ```

4. But this condition messes/changes the type I returned from TypeConverter and I always get `nonNull`.

----

Examples of annotation:

```php
// in Entity
// CmsApplicationStatus is enum

/**
 * @var string
 * @GraphQLType("CmsApplicationStatus")
 */
public $status;
```

```php
// in Dto
// CollectionTypeSettingsInput is custom type

/**
 * @var array
 * @GraphQLType("CollectionTypeSettingsInput")
 */
public $settings = [];
```

Reply:
This is really an edge case. Why not doing this:
```php
// in Entity
// CmsApplicationStatus is enum

/**
 * @var null|string
 * @GraphQLType("CmsApplicationStatus")
 */
public $status;
```

Reply:
Because it's a hack - I have to keep in mind that `@var` affects `@GraphQLType` and they must be used together. I also have cases where I don't have `@var` on property.

Ok, if it's an edge case then I will continue using my fork. Thanks